### PR TITLE
[DAF 4107] [IOS] pip show black screen when play from notification center

### PR DESCRIPTION
**Pull request by mvn-thanhluu-hn**
## Description
- Video change to black in pip mode when playing from notification center
- Add `video play output` for `player item`
  - Cannot find any reference now, I think when adding `video output` for `player item`, it makes `player` has high priority and still generates video frames when not visible

- I use function `add video output` from: https://developer.apple.com/library/archive/samplecode/AVGreenScreenPlayer/Listings/AVGreenScreenPlayer_GSPlayerView_m.html#//apple_ref/doc/uid/DTS40012325-AVGreenScreenPlayer_GSPlayerView_m-DontLinkElementID_10

## Ticket
https://dw-ml-nfc.atlassian.net/browse/DAF-4107

## Screenshot

- before:

https://github.com/dwango-nfc/betterplayer/assets/38061170/f39043b2-98dc-4d53-a395-cf40479ee94b

- after

https://github.com/dwango-nfc/betterplayer/assets/38061170/4175b2be-aae6-4450-ae5c-fbb42e3d65ee

### CODE REVIEW

File: example/lib/pages/picture_in_picture_page.dart [hunk 1]
```diff
@@ -50,12 +50,19 @@ class _PictureInPicturePageState extends State<PictureInPicturePage> {
           'betterPlayerEventType: ${event.betterPlayerEventType}, event.parameters: ${event.parameters.toString()}');
 
       if (event.betterPlayerEventType == BetterPlayerEventType.play) {
+        if (_willSwitchToPIPLayout) {
```
Comment (mvn-thanhluu-hn) [inline example/lib/pages/picture_in_picture_page.dart]: These changes in files are not related to PR, update for testing (pip close when pausing video)

Comment (HiroshiOshiro) [reply example/lib/pages/picture_in_picture_page.dart]: Do you need to this modification to be merge? 
Not sure why this modification was needed but If it is just a test and not needed in our app, I think should not be in PR.

**Review by HiroshiOshiro — CHANGES_REQUESTED** [review]
@mvn-thanhluu-hn 
Please check the comment.

File: example/lib/pages/picture_in_picture_page.dart (hunk 1 above)
Comment (mvn-thanhluu-hn) [reply example/lib/pages/picture_in_picture_page.dart]: Oh, each time I use this sample to test I always add this condition to test.

The case is in PIP mode if you press pause or play, app will close PIP (actually, it's implemented in `nfc` app with `isPIPMode` flag) so I think we should include it in PR.

Comment (mvn-thanhluu-hn) [reply example/lib/pages/picture_in_picture_page.dart]: It will affect android. I will update condition (only for iOS)

File: ios/Classes/BetterPlayer.m [hunk 2]
```diff
@@ -249,6 +249,13 @@ - (void)setDataSourcePlayerItem:(AVPlayerItem*)item withKey:(NSString*)key{
     _stalledCount = 0;
     _isStalledCheckStarted = false;
     _playerRate = 1;
+    
+    // Prevent pip change to black screen when play video from notification center
+    // When add video output to player item, maybe it makes player item high priority
+    // and player still generates frame when not visible
+    AVPlayerItemVideoOutput* playerItemVideoOutput = [[AVPlayerItemVideoOutput alloc] initWithPixelBufferAttributes:@{(id)kCVPixelBufferPixelFormatTypeKey: @(kCVPixelFormatType_32ARGB)}];
```
Comment (noboru-i) [inline ios/Classes/BetterPlayer.m:256]: Sorry.
How did you know the code?
If you have any reference, please add the link to PR description 🙏 

And I am afraid,
Now we don't have any "close" process?

- We call `addOutput`, so we don't need `removeOutput`?
- `playerItemVideoOutput` is "alloc"ed and "init"ed. We don't need to dispose it? (I'm afraid of memory leaks.)

**Review by noboru-i — COMMENTED** [review]
@mvn-thanhluu-hn 
It is difficult to understand it works...

I add a comment. Please check it 🙏

File: example/lib/pages/picture_in_picture_page.dart (hunk 1 above)
Comment (mvn-thanhluu-hn) [reply example/lib/pages/picture_in_picture_page.dart]: @HiroshiOshiro @noboru-i 
I updated the changes (refactoring names and handle like nfc app), these changes fix the bug pip close when pausing video.
Sorry if my expression makes you confused 🙏
https://github.com/dwango-nfc/betterplayer/pull/33/commits/cd5637c45171d97da0192dbe45cb0b5f527180a6

File: ios/Classes/BetterPlayer.m (hunk 2 above)
Comment (mvn-thanhluu-hn) [reply ios/Classes/BetterPlayer.m:256]: @noboru-i 

> If you have any reference, please add the link to PR description

I added it 🙏

> Now we don't have any "close" process?
> We call addOutput, so we don't need removeOutput?
> playerItemVideoOutput is "alloc"ed and "init"ed. We don't need to dispose it? (I'm afraid of memory leaks.)

Thank you for pointing out 👍

I checked again. In this case, only add `video output` for `player item` (not add any observers or other handling...) so it should be auto-released by IOS memory management system ([ARC](https://developer.apple.com/library/archive/documentation/Cocoa/Conceptual/MemoryMgmt/Articles/MemoryMgmt.html))
- `Player Item` is the only obj that contains `AVPlayerVideoOuput` (reference count = 1), so when `Player Item` is released -> `AVPlayerVideoOuput` should be released too.

For make sure, I use tool `Debug Memory Graph` on XCode and checked this case was not leaked
- Play, change data source -> only 1 obj in-memory tree
- Back to list screen -> can not found obj (obj is released)


Screen Shot 2023-07-10 at 15 15 59

Comment (noboru-i) [reply ios/Classes/BetterPlayer.m:256]: Thank you for adding and checking it detail.
I understand 👍

**Review by noboru-i — APPROVED** [review]
LGTM.

**Review by HiroshiOshiro — APPROVED** [review]
LGTM. Thanks.
